README.md
# CMS Sao

A microservice that orbits around [CMS](https://github.com/cms-dev/cms) exposing
the Entry, Result, Score and Scoreboard entities as [REST](https://en.wikipedia.org/wiki/Representational_state_transfer)
resources.

## Up and running

CMS Sao can be deployed and run as a Docker container or a standalone binary;
either way is recommend to run this application as the former one.

### Prerequisites

CMS Sao heavily relies on [CMS](https://github.com/cms-dev/cms) including its database, so
in order to have this application up and running you will need:

1. CMS 1.3.x or greater (the current Sao version was designed against the last CMS revision in Jan 2018)
2. CMS PostreSQL DB schema access (it could be the same credentials that CMS uses but is not recommended)
3. Docker engine 17.x or greater

### Deployment

CMS Sao can be deployed and run as a Docker container, it can be done directly
using `docker container run` command:

```shell
docker container run -p 8000:8000 jossemargt/cms-sao
```

Or it can be run using `docker-compose up` with a `docker-compose.yml` file
similar to the one on the project root.

### Configuration

All the intrinsic configurations can be overridden via the `config.properties`
file in Sao's working directory (you could use `config.properties.example` as
guide) or as environment variables with the `SAO_` prefix. As for example in
order to override the `datasource.host` value, you could start the Docker
container with the following syntax:

```shell
docker container run -p 8000:8000 -e 'SAO_CMS_DATASOURCE_HOST=10.10.37.10' jossemargt/cms-sao
```

When CMS Sao is executed using `docker-compose` the override values can be provided directly
in the `docker-compose.yml` or `docker-compose.override.yml` files (within the `environment`
section).

The most relevant properties are:

Property name | Default value | Description
--- | --- | ---
server.port | 8000 | The port where the application will listen for incoming requests
server.log.request | false | When set as true the application will log each incoming request to the STDOUT
server.error.tracedump | false | When set as true the application will include the error trace in failure responses
cms.url | http://localhost/ | The URL where Sao can stablish communication with CMS
cms.datasource.name | cmsdb | CMS PostgreSQL schema name
cms.datasource.username | cmsuser | CMS PostgreSQL datasource username
cms.datasource.password | notsecure | CMS PostgreSQL datasource password
cms.datasource.host | 127.0.0.1 | CMS PostgreSQL host network address
cms.datasource.port | 5432 | CMS PostgreSQL instance port
cms.datasource.sslmode | disable | CMS PostgreSQL [ssl mode](https://www.postgresql.org/docs/9.1/libpq-ssl.html) (valid values: require, disable, verify-ca, verify-full)

## License

This project is licensed under the **MIT License** - see the [LICENSE](LICENSE)
file for details.


CHANGELOG.md
# Changelog

## 0.5.0 (next)

- Enable tooling to generate code using vendor dependencies.

## 0.4.0

- Expose Draft Result resource (analogous to CMS UserTestResult entity)
- Fix Entry and Draft list endpoints to display unprocessed resources

## 0.3.0

- Expose Result resource (analogous to CMS SubmissionResult entity)
- Correctly determine the `token` value in Entry resource
- Add `language` field to Entry and Draft resources
- Package CMS Sao binary as Docker image

## 0.2.0

- Expose Draft resource (analogous to CMS UserTest entity)

## 0.1.0

- Expose Entry resource (analogous to CMS Submission entity)


main.go
//go:generate go run gen/main.go

package main

import (
	"fmt"
	"net/http"
	"time"

	"github.com/goadesign/goa"
	"github.com/goadesign/goa/middleware"
	"github.com/hashicorp/go-retryablehttp"
	"github.com/jmoiron/sqlx"
	_ "github.com/lib/pq"
	goup "github.com/ufoscout/go-up"

	"github.com/jossemargt/cms-sao/app"
	"github.com/jossemargt/cms-sao/storage"
)

func main() {
	// Resolve configurations
	var up goup.GoUp
	{
		up, _ = goup.NewGoUp().
			// Read default config file if any
			AddFile("./config.properties", true).
			// Read environment variables with prefix SAO_
			AddReader(goup.NewEnvReader("SAO_", true, true)).
			Build()
	}

	var httpClient *httpRetryableClient
	{
		httpClient = new(httpRetryableClient)
		retryable := retryablehttp.NewClient()

		// Max retries
		retryable.RetryMax = up.GetIntOrDefault("cms.http.retry.limit", 5)

		// Minimal retry wait (in seconds)
		minWait := up.GetIntOrDefault("cms.http.retry.waitmin", 1)
		retryable.RetryWaitMin = time.Duration(minWait) * time.Second

		// Max retry wait
		maxWait := up.GetIntOrDefault("cms.http.retry.waitmax", 30)
		retryable.RetryWaitMax = time.Duration(maxWait) * time.Second

		httpClient.retryable = retryable
	}

	var dbConn *sqlx.DB
	{
		dbhost := up.GetStringOrDefault("cms.datasource.host", "localhost")
		dbport := up.GetStringOrDefault("cms.datasource.port", "5432")
		dbname := up.GetStringOrDefault("cms.datasource.name", "cmsdb")
		dbuser := up.GetStringOrDefault("cms.datasource.username", "cmsuser")
		dbpassword := up.GetStringOrDefault("cms.datasource.password", "")
		dbsslmode := up.GetStringOrDefault("cms.datasource.sslmode", "require")

		dbConn = sqlx.MustConnect("postgres",
			fmt.Sprintf("user=%s password=%s dbname=%s host=%s port=%s sslmode=%s",
				dbuser,
				dbpassword,
				dbname,
				dbhost,
				dbport,
				dbsslmode,
			),
		)
	}

	// Create resource repositories
	entryRepository := storage.NewEntryRepository(dbConn)
	resultRepository := storage.NewResultRepository(dbConn)
	draftRepository := storage.NewEntryDraftRepository(dbConn)
	draftresultRepository := storage.NewDraftResultRepository(dbConn)

	// Create service
	service := goa.New("SAO v1")

	// Mount middleware
	service.Use(middleware.RequestID())
	service.Use(middleware.LogRequest(up.GetBoolOrDefault("server.log.request", false)))
	service.Use(middleware.ErrorHandler(service, up.GetBoolOrDefault("server.response.tracedump", false)))
	service.Use(middleware.Recover())

	// Mount "actions" controller
	c := NewActionsController(service)
	app.MountActionsController(service, c)
	// Mount "draft" controller
	c2 := NewDraftController(service, draftRepository)
	app.MountDraftController(service, c2)
	// Mount "draft-result" controller
	c3 := NewDraftResultController(service, draftresultRepository)
	app.MountDraftresultController(service, c3)
	// Mount "entry" controller
	c4 := NewEntryController(service, entryRepository)
	app.MountEntryController(service, c4)
	// Mount "result" controller
	c5 := NewResultController(service, resultRepository)
	app.MountResultController(service, c5)
	// Mount "submit entry transaction" controller
	c6 := NewEntrySubmitTrxController(service)
	app.MountEntrySubmitTrxController(service, c6)
	// Mount "submit entry draft transaction" controller
	c7 := NewDraftSubmitTrxController(service)
	app.MountDraftSubmitTrxController(service, c7)

	// Start service
	serverAddr := fmt.Sprintf("%s:%s", "", up.GetStringOrDefault("server.port", "8000"))
	if err := service.ListenAndServe(serverAddr); err != nil {
		service.LogError("startup", "err", err)
	}

}

type httpRetryableClient struct {
	retryable *retryablehttp.Client
}

func (c *httpRetryableClient) Do(req *http.Request) (*http.Response, error) {
	request, err := retryablehttp.NewRequest(req.Method, req.URL.String(), req.Body)

	if err != nil {
		return nil, err
	}

	return c.retryable.Do(request)
}
